platforms/allwinner-d1/drivers/src/plic.rs

```diff
@@ -0,0 +1,287 @@
+use core::{
+    ptr::null_mut,
+    sync::atomic::{AtomicPtr, Ordering},
+};
+
+use d1_pac::{plic, Interrupt, PLIC};
+
+/// Interrupt Priority from 0..31
+pub type Priority = plic::prio::PRIORITY_A;
+
+#[doc = r" TryFromPrioritytError"]
+#[derive(Debug, Copy, Clone)]
+pub struct TryFromPriorityError(());
+
+/// Platform-Level Interrupt Controller (PLIC) interface
+pub struct Plic {
+    plic: PLIC,
+}
+
+impl Plic {
+    /// Create a new `Plic` from the [`PLIC`](d1_pac::PLIC) peripheral
+    pub fn new(plic: PLIC) -> Self {
+        // TODO any initial setup we should be doing for the PLIC at startup?
+        Self { plic }
+    }
+
+    /// Obtain a static `Plic` instance for use in e.g. interrupt handlers
+    ///
+    /// # Safety
+    ///
+    /// 'Tis thine responsibility, that which thou doth summon.
+    pub unsafe fn summon() -> Self {
+        Self {
+            plic: d1_pac::Peripherals::steal().PLIC,
+        }
+    }
+
+    /// Enable an interrupt
+    ///
+    /// # Safety
+    ///
+    /// May effect normal interrupt processing
+    pub unsafe fn unmask(&self, interrupt: Interrupt) {
+        let nr = interrupt.into_bits() as usize;
+        let (reg_offset, irq_en) = (nr / 32, 1 << (nr % 32));
+        self.plic.mie[reg_offset].modify(|r, w| w.bits(r.bits() | irq_en));
+    }
+
+    /// Disable an interrupt
+    pub fn mask(&self, interrupt: Interrupt) {
+        let nr = interrupt.into_bits() as usize;
+        let (reg_offset, irq_en) = (nr / 32, 1 << (nr % 32));
+        self.plic.mie[reg_offset].modify(|r, w| unsafe { w.bits(r.bits() & !irq_en) });
+    }
+
+    /// Globally set priority for one interrupt
+    ///
+    /// # Safety
+    ///
+    /// May effect normal interrupt processing
+    pub unsafe fn set_priority(&self, interrupt: Interrupt, priority: Priority) {
+        let nr = interrupt.into_bits() as usize;
+        self.plic.prio[nr].write(|w| w.bits(priority.into_bits()));
+    }
+
+    pub fn claim(&self) -> Interrupt {
+        let claim = self.plic.mclaim.read().mclaim().bits() as u8;
+        match Interrupt::try_from(claim) {
+            Ok(interrupt) => interrupt,
+            Err(_) => {
+                panic!("error claiming interrupt");
+            }
+        }
+    }
+
+    pub fn complete(&self, interrupt: Interrupt) {
+        self.plic
+            .mclaim
+            .write(|w| w.mclaim().variant(interrupt.into_bits() as u16));
+    }
+
+    pub unsafe fn register(&self, interrupt: Interrupt, new_hdl: fn()) {
+        let v = INTERRUPT_ARRAY.iter().find(|v| v.id == interrupt as u16);
```

Comment:
are the IDs numerically increasing, and if so, can we just put each interrupt at its' ID's index in the array, instead of doing this? not a blocker, just curious.

Reply:
We could, we would just need a sparse array, as some interrupt IDs are not used.

I think there is SOME way to get the hardware to do normal ISR dispatching, instead of a single "hardware interrupt method", but I haven't looked up how to do that yet.

We also could probably binary search, but I decided to go with 'bad and simple and done' for now.